crates/store/re_log_encoding/src/codec/file/decoder.rs

```diff
@@ -0,0 +1,86 @@
+use super::{MessageHeader, MessageKind};
+use crate::codec::arrow::decode_arrow;
+use crate::codec::CodecError;
+use crate::decoder::DecodeError;
+use re_log_types::LogMsg;
+use re_protos::TypeConversionError;
+
+impl MessageKind {
+    pub(crate) fn decode(data: &mut impl std::io::Read) -> Result<Self, DecodeError> {
+        let mut buf = [0; 4];
+        data.read_exact(&mut buf)?;
+
+        match u32::from_le_bytes(buf) {
+            1 => Ok(Self::SetStoreInfo),
+            2 => Ok(Self::ArrowMsg),
+            3 => Ok(Self::BlueprintActivationCommand),
+            255 => Ok(Self::End),
+            _ => Err(DecodeError::Codec(CodecError::UnknownMessageHeader)),
+        }
+    }
+}
+
+impl MessageHeader {
+    pub(crate) fn decode(data: &mut impl std::io::Read) -> Result<Self, DecodeError> {
+        let kind = MessageKind::decode(data)?;
+        let mut buf = [0; 4];
+        data.read_exact(&mut buf)?;
+        let len = u32::from_le_bytes(buf);
+
+        Ok(Self { kind, len })
+    }
+}
+
+pub(crate) fn decode(data: &mut impl std::io::Read) -> Result<(u64, Option<LogMsg>), DecodeError> {
+    use re_protos::external::prost::Message;
+    use re_protos::log_msg::v0::{ArrowMsg, BlueprintActivationCommand, Encoding, SetStoreInfo};
+
+    let mut read_bytes = 0u64;
+    let header = MessageHeader::decode(data)?;
+    read_bytes += std::mem::size_of::<MessageHeader>() as u64 + header.len as u64;
+
+    let mut buf = vec![0; header.len as usize];
+    data.read_exact(&mut buf[..])?;
+
+    let msg = match header.kind {
+        MessageKind::SetStoreInfo => {
+            let set_store_info = SetStoreInfo::decode(&buf[..])?;
+            Some(LogMsg::SetStoreInfo(set_store_info.try_into()?))
+        }
+        MessageKind::ArrowMsg => {
+            let arrow_msg = ArrowMsg::decode(&buf[..])?;
+            if arrow_msg.encoding() != Encoding::ArrowIpc {
+                return Err(DecodeError::Codec(CodecError::UnsupportedEncoding));
+            }
+
+            let (schema, chunk) = decode_arrow(
+                &arrow_msg.payload,
+                arrow_msg.uncompressed_size as usize,
+                arrow_msg.compression().into(),
+            )?;
+
+            let store_id: re_log_types::StoreId = arrow_msg
+                .store_id
+                .ok_or_else(|| {
+                    TypeConversionError::missing_field("rerun.log_msg.v0.ArrowMsg", "store_id")
```

Comment:
Where does the `v0` come from? Is there a way not to hard-code it here?

Reply:
It's part of package name, I wanted to be explicit, because while debugging this I was bitten by it not being explicit enough.
It turns out that `prost_build` can generate an impl of `prost::Name` which has these strings statically defined (but weirdly enough `tonic_build` does not expose that option, even though it works), so I made a macro that accepts the type and field name instead, and uses the constants from `prost::Name` in the error message.

Reply:
I wish we could auto-generate this code instead, that would be a lot better, but it seems difficult unless the types map 1:1 exactly.